### PR TITLE
Fiddle with VIC parser and new data

### PR DESCRIPTION
**Pull request by s3364859**
@6nop @s3364942 @s3494110 @s3482247 

In this PR:
 - Fiddled about with the code getting changes to the VIC KML data to play nicely

reverse

### CODE REVIEW

File: CPT331.Data.Parsers/VicKmlParser.cs [hunk 1]
```diff
@@ -22,49 +22,25 @@ public VicKmlParser(string dataSourceDirectory)
 
 		internal const string VIC = "VIC";
 
-		private static Dictionary<string, string> CreateMappingDictionary(string fileName)
```
Comment (s3364859) [inline CPT331.Data.Parsers/VicKmlParser.cs]: No longer needed.

File: CPT331.Data.Parsers/VicKmlParser.cs [hunk 2]
```diff
@@ -22,49 +22,25 @@ public VicKmlParser(string dataSourceDirectory)
...
 		protected override void OnParse(string fileName, List<Coordinate> coordinates)
 		{
 			OutputStreams.WriteLine($"Parsing {VIC} data...");
 
-			Dictionary<string, string> mappingDictionary = CreateMappingDictionary(fileName);
-
 			XmlDocument xmlDocument = new XmlDocument();
 			xmlDocument.Load(fileName);
 
+			string[] reverseLocalGovernmentAreaNames = new string[]
```
Comment (s3364859) [inline CPT331.Data.Parsers/VicKmlParser.cs]: These have their windings reversed.

File: CPT331.Data.Parsers/VicKmlParser.cs [hunk 3]
```diff
@@ -84,7 +60,12 @@ protected override void OnParse(string fileName, List<Coordinate> coordinates)
 				
 					coordinates.Add(new Coordinate(Double.Parse(coordinateParts[1]), Double.Parse(coordinateParts[0])));
 				}
-				
+
+				if (reverseLocalGovernmentAreaNames.Contains(name) == false)
```
Comment (s3364859) [inline CPT331.Data.Parsers/VicKmlParser.cs]: The get reversed here!

File: CPT331.Core/ObjectModel/Coordinate.cs [hunk 4]
```diff
@@ -11,10 +12,16 @@ public class Coordinate
 		public Coordinate(double latitude, double longitude)
 		{
             ValidateCordinate(latitude, longitude);
-            _latitude = latitude;
+
+			_latitude = latitude;
 			_longitude = longitude;
 		}
 
+		public const double LatitudeMaxValue = 90;
```
Comment (s3364859) [inline CPT331.Core/ObjectModel/Coordinate.cs]: I had some time to really over think this one!

File: CPT331.Data.Migration/Options.cs [hunk 5]
```diff
@@ -13,9 +13,9 @@ internal class Options
 	{
 		private string _drop;
 		private string _kml;
-		private string _xml;
-        private bool _simpleRecoveryModel;
 		private IParserState _lastParserState;
+		private bool _simpleRecoveryModel;
```
Comment (s3364859) [inline CPT331.Data.Migration/Options.cs]: Reordered to satisfy my OCD.

File: CPT331.Data.Migration/Options.cs [hunk 6]
```diff
@@ -43,42 +43,42 @@ public string Kml
 			}
 		}
 
-		[Option('x', "xml", Required = false, HelpText = "Process the XML data sources. Use this option to hydrate an empty database.")]
-		public string Xml
+		[ParserState]
+		public IParserState LastParserState
```
Comment (s3364859) [inline CPT331.Data.Migration/Options.cs]: OCD.

File: CPT331.WebAPI/Controllers/CrimeController.cs [hunk 7]
```diff
@@ -48,18 +46,11 @@ public CrimeModel Crime(int id)
 
 		[HttpGet]
 		[Route("api/Crime/CrimesByCoordinate")]
+		[ValidateCoordinates]
 		public CrimeByCoordinateModel CrimesByCoordinate(double latitude, double longitude, int count = DefaultNumberOfCrimeRecords, string sortBy = "", SortDirection? sortDirection = null)
 		{
-            CrimeByCoordinateModel crimeByCoordinateModel = null;
```
Comment (s3364859) [inline CPT331.WebAPI/Controllers/CrimeController.cs]: All this validation is now moved into the appropriate action attribute.

File: CPT331.WebAPI/Validation/ValidateCoordinatesAttribute.cs [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+using System.Text;
+using System.Web.Http.Controllers;
+using System.Web.Http.Filters;
+using System.Web.Http.ModelBinding;
+
+#endregion
+
+namespace CPT331.WebAPI.Validation
+{
+	public class ValidateCoordinatesAttribute : ActionFilterAttribute
```
Comment (s3364859) [inline CPT331.WebAPI/Validation/ValidateCoordinatesAttribute.cs]: Coordinates are validated by this action attribute - requests won't even make it to the controller unless they pass this first.

File: CPT331.WebAPI/Validation/ValidateCoordinatesAttribute.cs [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+						stringBuilder.AppendLine(((modelError.Exception == null) ? modelError.ErrorMessage : modelError.Exception.ToString()));
+					}
+				}
+
+				HttpResponseMessage httpResponseMessage = new HttpResponseMessage(HttpStatusCode.BadRequest)
+				{
+					Content = new StringContent(stringBuilder.ToString())
+				};
+
+				httpActionContext.Response = httpResponseMessage;
```
Comment (s3364859) [inline CPT331.WebAPI/Validation/ValidateCoordinatesAttribute.cs]: No controller for you!

File: README.md [hunk 10]
```diff
@@ -1,6 +1,123 @@
-# CPT331
-## Totally cool project for Cool People Only (TM)
-Write some interesting stuff here.
+# EventGuardian
```
Comment (s3364859) [inline README.md]: Completely updated readme.

File: README.md [hunk 11]
```diff
@@ -1,6 +1,123 @@
...
+Contains repository types.
+
+### CPT331.Data.Migration
+Console app used to create and populate the database.
+
+### CPT331.Data.Parsers
+Contains parser types used to convert various file-based data sources into a common object model.
+
+### CPT331.iOS
+iOS app implementation.
```
Comment (s3364859) [inline README.md]: @s3482247 Mate, you'll need to flesh this bit out for us.

File: CPT331.WebAPI/Validation/ValidateCoordinatesAttribute.cs [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+
+				if (Double.TryParse(longitudeModelState.Value.AttemptedValue, out longitude) == false)
+				{
+					longitudeModelState.Errors.Add(new FormatException($"Longitude value of '{longitude}' could not be converted."));
+				}
+			}
+
+			if ((latitude != 0) && (longitude != 0))
+			{
+				if (Coordinate.IsLatitudeValid(latitude) == false)
```
Comment (s3364859) [inline CPT331.WebAPI/Validation/ValidateCoordinatesAttribute.cs]: Actual validation relies on the Coordinate type.

File: CPT331.WebAPI/Controllers/CrimeController.cs (hunk 7 above)
Comment (6nop) [inline CPT331.WebAPI/Controllers/CrimeController.cs]: Nice! I like it.

File: CPT331.Web/Models/Admin/StateModel.cs [hunk 13]
```diff
@@ -52,7 +55,7 @@ public string AbbreviatedName
 			}
 		}
 
-		[DataType(DataType.Text)]
+		[DataType(DataType.DateTime)]
```
Comment (s3364859) [inline CPT331.Web/Models/Admin/StateModel.cs]: Multiple whoops' here.

File: CPT331.WebAPI.Parsers/EventFindaWebParser.cs [hunk 14]
```diff
@@ -43,7 +43,7 @@ public List<EventInfo> GetEventsByCoordinate(double latitude, double longitude,
 		{
 			Dictionary<string, string> queryString = new Dictionary<string, string>();
 			queryString.Add("offset", Offset.ToString());
-			queryString.Add("point", $"{latitude},{longitude}");
+			queryString.Add("point", $"{longitude},{latitude}");
```
Comment (s3364859) [inline CPT331.WebAPI.Parsers/EventFindaWebParser.cs]: Insert the appropriate face here.

Comment (6nop) [inline CPT331.WebAPI.Parsers/EventFindaWebParser.cs]: I think you were right the first time!
/uppercut?

**Comment by 6nop** [comment]
All good now 👍

**Pull request merged**